Fixes the total num of payments returned by QueryPayments when there are dates restrictions

MPins opened the pull request:
Fixes #8530

## Change Description
The function QueryPayments was not considering the dates restriction when counting the total num of payments.

## Steps to Test
`lncli listpayments --creation_date_start 1650000000 --count_total_payments --creation_date_end 1650000001 | grep total_num_payments`

This command should return 0 as there are no payments between the date start and the date end.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -597,28 +597,7 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
 		// literally have to traverse the cursor linearly, which can
 		// take quite a while. So it's an optional query parameter.
 		if query.CountTotal {
```
can we update the list payments itest to check this new behavior?

MPins replied:
Yes, good idea. I'm going to work on it on the following days.

yyforyongyu reviewed (changes requested):
Thanks for the PR. Please check the contribution guidelines regarding [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure).

lightninglabs-deploy commented:
@yyforyongyu: review reminder

yyforyongyu commented on itest/lnd_payment_test.go:
```diff
@@ -273,13 +281,15 @@ func testListPayments(ht *lntest.HarnessTest) {
 			resp := alice.RPC.ListPayments(req)
 
 			if tc.expected {
 				require.Lenf(t, resp.Payments, 1, "req=%v", req)
```
also check `resp.TotalNumPayments` is 1 here?

MPins replied:
Yes, you right. It makes the test more complete. I'll be changing it.

MPins replied:
done 07bf0a1968280138fbc5496cac3774bdbb89e82e

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -597,28 +597,7 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
 		// literally have to traverse the cursor linearly, which can
```
typo in commit msg: toyal -> total

MPins replied:
done 6fecd6211c17aa7fa7e7b7bd4911daab7bbf5f3e

yyforyongyu commented on itest/lnd_payment_test.go:
```diff
@@ -256,6 +256,14 @@ func testListPayments(ht *lntest.HarnessTest) {
 				endDate:  createTimeSeconds - 1,
 				expected: false,
 			},
+			{
```
think we can remove the line `modified:   itest/lnd_payment_test.go` in the commit msg

MPins replied:
done 07bf0a1968280138fbc5496cac3774bdbb89e82e

yyforyongyu reviewed (changes requested):
Approved CI run.

feelancer21 commented:
Imho, this isn't a bug. It's a feature.

https://github.com/lightningnetwork/lnd/blob/a9655357ca697a01997b941b16be31db611cbf5c/lnrpc/lightning.proto#L4211-L4217

feelancer21 commented:
@MPins I don't think your fix solves the issue. Using `len` caps the number of payments at `max_payment`, which contradicts the definition of the flag `count_total_payments`. You need to count all payments that are in the database and within the specified date range.

https://github.com/lightningnetwork/lnd/blob/a9655357ca697a01997b941b16be31db611cbf5c/lnrpc/lightning.proto#L4178-L4184

MPins commented:
> independent of the 
>      max_payments parameter

At this point, the offset was exhausted and all invoices, that respect the query, were collected to 'resp.Payments', so 'uint64(len(resp.Payments))' will return that total num of payments.

feelancer21 commented:
> > independent of the
> > max_payments parameter
> 
> At this point, the offset was exhausted and all invoices, that respect the query, were collected to `resp.Payments`, so `uint64(len(resp.Payments))` will return that total num of payments.

I am also beginner. As far I understand the code, the maximum length of `res.Payments` is restricted to `MaxPayments` because of the paginator. And hence it is not independent of `max_payments`.

MPins commented:
> > > independent of the
> > > max_payments parameter
> > 
> > 
> > At this point, the offset was exhausted and all invoices, that respect the query, were collected to `resp.Payments`, so `uint64(len(resp.Payments))` will return that total num of payments.
> 
> I am also beginner. As far I understand the code, the maximum length of `res.Payments` is restricted to `MaxPayments` because of the paginator. And hence it is not independent of `max_payments`.
> 
> Edit: I made a test on testnet `lncli listpayments --count_total_payments --max_payments 1`. And `total_num_payments` changed to 1 now. I think it is not as expected.

You right! Thanks! I'll be working on it.

MPins commented:
> > > > independent of the
> > > > max_payments parameter
> > > 
> > > 
> > > At this point, the offset was exhausted and all invoices, that respect the query, were collected to `resp.Payments`, so `uint64(len(resp.Payments))` will return that total num of payments.
> > 
> > 
> > I am also beginner. As far I understand the code, the maximum length of `res.Payments` is restricted to `MaxPayments` because of the paginator. And hence it is not independent of `max_payments`.
> > Edit: I made a test on testnet `lncli listpayments --count_total_payments --max_payments 1`. And `total_num_payments` changed to 1 now. I think it is not as expected.
> 
> You right! Thanks! I'll be working on it.

@yyforyongyu now the correction is considering the point indicate by @feelancer21, when the request comes with --count_total_payments and --max_payments causing pagination. Ready to go, thank you both!

feelancer21 commented:
Thanks. I see some redundant code now, reading the payment and checking the criteria are mainly the same in `accumulatePayments` and `countFn`. You should check if it is possible to extract this part in a seperate functions processing the steps of reading and checking.

MPins committed:
channeldb: Fixes the count of payments returned by QueryPayments

QueryPayments function was changed to return the total num of payments
considering the date start and date end restrictions existing on the query.

MPins committed:
itest: Checking the total num of payments

When there are no payments expected returning, besides
checking if the paymensts is empty, check also if the
total num of payments is zero.

MPins commented:
> Thanks. I see some redundant code now, reading the payment and checking the criteria are mainly the same in `accumulatePayments` and `countFn`. You should check if it is possible to extract this part in a seperate functions processing the steps of reading and checking.

Yes, thats make sense. Done!

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -529,6 +529,18 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
 		}
 
+		// Helper function to check if a payment meets the creation time criteria.
+		paymentinRange := func(payment *MPPayment) bool {
```
```suggestion
		paymentInRange := func(payment *MPPayment) bool {
```

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -557,21 +569,8 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
-				createTime > query.CreationDateEnd {
-
+			// Check if the payment is within the time range.
+			if !paymentinRange(payment) {
```
```suggestion
			if !paymentInRange(payment) {
```

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -601,9 +600,27 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
+					return err
+				}
+				// Check if the payment is within the time range.
+				if !paymentinRange(payment) {
```
```suggestion
				if !paymentInRange(payment) {
```

yyforyongyu reviewed (changes requested):
Linter failed, and a nit in the method name. Could you add a release note to 0.18.3? Then it should be good to go.

MPins commented:
> Linter failed, and a nit in the method name. Could you add a release note to 0.18.3? Then it should be good to go.

Thanks!

The linter failed because there was a missing blank line before a 'return'. I can see that in others parts of the code, is there an specific case where the blank line is a must?

The release note could be the following:
* [Fixed a bug](https://github.com/lightningnetwork/lnd/pull/8883) when RPC
  call is made to list payments counting the total number of payments.

Should I update the release note file, or you do?

yyforyongyu commented:
> The linter failed because there was a missing blank line before a 'return'. I can see that in others parts of the code, is there an specific case where the blank line is a must?

It's needed for readability. The linter only checks new code based on a configured look-ahead commit hash. That's probably why you'd see some other parts still use this pattern.
 
> Should I update the release note file, or you do?

Yeah it's usually the author who does the update. You'll find the right section and add the note, also update the author list in the end of the release note file.

yyforyongyu commented:
Approved CI run

MPins committed:
docs: release-notes-0.18.3

MPins commented:
@yyforyongyu just commited the release notes 0.18.3 changes. Thank you!

MPins commented:
@yyforyongyu I did not understand the lint fail!! Any Tip?

guggero commented on channeldb/payments.go:
```diff
@@ -529,6 +529,23 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
+		// criteria.
+		paymentInRange := func(payment *MPPayment) bool {
+			createTime := payment.Info.CreationTime.Unix()
+			if createTime < query.CreationDateStart {
```
There shouldn't be an empty line after this `if` statement:
```
channeldb/payments.go:536: unnecessary leading newline (whitespace)
```

The extra line is only needed if the condition in the `if` statement spans over multiple lines, such as is the case in the next one. So the extra newline is a visual helper to show that a condition is multi-line.

MPins replied:
Understood. Thanks. Done.

yyforyongyu commented:
hmm this has failed in two of the itests,
```
--- FAIL: TestLightningNetworkDaemon/tranche02/45-of-159/bitcoind/list_payments (22.88s)
        --- FAIL: TestLightningNetworkDaemon/tranche02/45-of-159/bitcoind/list_payments/payment_exact_start_date (0.01s)

harness_rpc.go:97: 
        	Error Trace:	/home/runner/work/lnd/lnd/lntest/rpc/harness_rpc.go:97
        	            				/home/runner/work/lnd/lnd/lntest/rpc/lnd.go:187
        	            				/home/runner/work/lnd/lnd/itest/lnd_payment_test.go:286
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unknown desc = error rolling back tx: conn busy
        	Messages:   	Alice: failed to call ListPayments
```

Think it's related?

MPins commented:
> list_payments/payment_exact_start_date

I had run the tests after the change and they are running properly!

 Running integration tests with btcd backend.
rm -rf itest/*.log itest/.logs-*; date
qua 31 jul 2024 18:45:10 -03
EXEC_SUFFIX= scripts/itest_part.sh 0 1 -test.run="TestLightningNetworkDaemon/tranche.*/.*-of-.*/.*/list_payments" -test.timeout=180m 
/home/pins/Projects/lnd/itest/itest.test -test.v -test.run=TestLightningNetworkDaemon/tranche.*/.*-of-.*/.*/list_payments -test.timeout=180m -logoutput -logdir=.logs-tranche0 -lndexec=/home/pins/Projects/lnd/itest/lnd-itest -btcdexec=/home/pins/Projects/lnd/itest/btcd-itest -splittranches=1 -runtranche=0
=== RUN   TestLightningNetworkDaemon
    harness_setup.go:25: Setting up HarnessTest...
    harness_setup.go:38: Prepare the miner and mine blocks to activate segwit...
    harness_setup.go:46: Connecting the miner at 127.0.0.1:10446 with the chain backend...
    harness.go:313: Setting up standby nodes Alice and Bob...
    harness.go:352: Finished the setup, now running tests...
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_exact_start_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_earlier_start_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_future_start_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_exact_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_future_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_earlier_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_earlier_start_and_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_exact_start_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_earlier_start_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_future_start_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_exact_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_future_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_earlier_end_date
=== RUN   TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_earlier_start_and_end_date
=== NAME  TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments
    harness.go:440: finished test: list_payments, start height=440, end height=447, mined blocks=7
=== NAME  TestLightningNetworkDaemon
    lnd_test.go:155: =========> tests finished for tranche: 0, tested 156 cases, end height: 440
--- PASS: TestLightningNetworkDaemon (90.83s)
    --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments (19.19s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_exact_start_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_earlier_start_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_future_start_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_exact_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_future_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_earlier_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/payment_earlier_start_and_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_exact_start_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_earlier_start_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_future_start_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_exact_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_future_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_earlier_end_date (0.00s)
        --- PASS: TestLightningNetworkDaemon/tranche00/45-of-156/btcd/list_payments/invoice_earlier_start_and_end_date (0.00s)
PASS

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -601,9 +604,28 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
-			countFn := func(_, _ []byte) error {
-				totalPayments++
 
+			countFn := func(sequenceKey, hash []byte) error {
```
This func won't work in SQL unless we change `ForAll` to `ForEach` below.

MPins replied:
I did the change ... now it is working even to SQL databases ... but as you said there is no otimization when there is a SQL database available because we cannot use the `ForAll` instead of `ForEach` as we done before this fix.

As you said below maybe we shouldn't touch it until we can find an efficient counting method. Maybe this is another PR to come!

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -601,9 +604,28 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
 
+			countFn := func(sequenceKey, hash []byte) error {
+				r := bytes.NewReader(hash)
+				paymentHash, err := deserializePaymentIndex(r)
```
Since counting payments is already slow, this would make it even slower - for nodes with millions of payments this would take a very long time. I think at least we should make this query conditional, that,
1. if there aren't any date restrictions, use the original `totalPayments++` approach.
2. if there is a date range, use this `countFn`.

But I still think this would be very slow, cc @bhandras for some help here, maybe we shouldn't touch it until we can find an efficient counting method?

MPins replied:
Maybe there is a way to measure the performance impact and only move forward if it is near the same of we have today without the fix.

yyforyongyu reviewed (commented):
To see the failure, you need to run it against the SQL db via `make itest icase=list_payments dbbackend=postgres`.

Also I'm not sure if #8530 is indeed an issue - what if a user actually wants to know the total num of payments instead of the ranged payments?

If we want to go with the current design, we should also update the `count_total_payments` here to be `count_payments`, and explain in details when the total num will be counted, and when the ranged num will be counted,
https://github.com/lightningnetwork/lnd/blob/04dde98edcc0a85699986374ef950687cd043871/lnrpc/lightning.proto#L4184-L4190

MPins commented:
> sue - what if a user actually wants to know the total num of payments instead of the ranged payments?

About it is an issue or not, I have the following opinion:
1 - If the user want the total num of payments with no date restrictions, why user would set the date range? It is counter intuitive.
2 - Without this fix there is no way to the user know the total num of payments in a date range.
3 - In the text you highlighted it says "independent of the max_payments parameter" not independent of date range.

lightninglabs-deploy commented:
@guggero: review reminder
@mpins, remember to re-request review from reviewers when ready

guggero commented:
I'm not sure how to proceed here...

It sounds to me like the expectation for what the `--count_total_payments` actually does is different between the author of the issue and the reviewers.

Maybe what we need is a new flag, `--count_payments_in_range` that does the new count as implemented currently but warns the user about the potential performance impact.

Then we can change the description of the `--count_total_payments` flag to say "counts all payments in the database independent of any filter/range criteria given".

yyforyongyu commented:
Thanks a lot for the input so far @MPins ! I'd recommend pausing the development of this PR at the moment as we will soon bring native SQL support for payments in 0.19.0 - maybe we can find a more efficient way to perform the queries by then.

lightninglabs-deploy commented:
@mpins, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mpins, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mpins, remember to re-request review from reviewers when ready

yyforyongyu commented:
!lightninglabs-deploy mute

saubyk commented:
> @bhandras @yyforyongyu and @saubyk - do you think it’s a good time to resume work on this PR?

don't think so. Our plans for the sql schema for payments moved to 0.20, so it would be best to work on this after

Bartok9 commented:
## Review note

Thanks for the early fix attempt on #8530.

This PR is now **CONFLICTING** against current `master` (payments live under `payments/db`, not the old `channeldb` paths this diff touches), and checklist/tests/release-notes target are stale relative to current tree.

Prefer [#10874](https://github.com/lightningnetwork/lnd/pull/10874), which:
- fixes **both** KV and SQL backends on current layout
- adds regression tests (including mixed in-flight / succeeded semantics)
- includes a 0.22 release note

Suggest closing this as superseded by #10874 once maintainers agree, or fully rebasing onto current `payments/db` if you still want a competing approach (likely more work than helping review #10874).

MPins commented:
This bug will be addressed in #10874 .
